unblob/handlers/filesystem/romfs.py

```diff
@@ -35,6 +67,291 @@ def get_string(file: io.BufferedIOBase) -> bytes:
+        if not is_safe_path(extract_root, inode.path):
+            logger.warn("Path traversal attempt, discarding.", output_path=output_path)
+            return
+        logger.info("dumping inode", inode=inode, output_path=str(output_path))
+
+        if inode.type == FS_TYPE.HARD_LINK:
+            self.create_hardlink(extract_root, output_path, inode)
+        elif inode.type == FS_TYPE.SYMLINK:
+            self.create_symlink(extract_root, output_path, inode)
+        elif inode.type == FS_TYPE.DIRECTORY:
+            output_path.mkdir(mode=inode.mode, exist_ok=True)
+        elif inode.type == FS_TYPE.FILE:
+            with output_path.open("wb") as f:
+                f.write(inode.content)
+        elif inode.type in [FS_TYPE.BLOCK_DEV, FS_TYPE.CHAR_DEV]:
+            os.mknod(inode.path, inode.mode, inode.dev)
+        elif inode.type == FS_TYPE.FIFO:
+            os.mkfifo(output_path, inode.mode)
+
+    def dump_fs(self):
+        # first we create files and directories
+        fd_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.FILE, FS_TYPE.DIRECTORY, FS_TYPE.FIFO, FS_TYPE.SOCKET]
+        }
+        for inode in sorted(fd_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+        if os.geteuid() != 0:
+            logger.warn(
+                "root privileges are required to create block and char devices, skipping."
+            )
+        else:
+            # then we create devices if we have enough privileges
+            dev_inodes = {
+                k: v
+                for k, v in self.inodes.items()
+                if v.type in [FS_TYPE.BLOCK_DEV, FS_TYPE.CHAR_DEV]
+            }
+            for inode in sorted(dev_inodes.values(), key=lambda inode: inode.path):
+                self.create_inode(self.extract_root, inode)
+
+        # then we create links
+        links_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.SYMLINK, FS_TYPE.HARD_LINK]
+        }
+        for inode in sorted(links_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+    def __str__(self):
+        return f"signature: {self.signature}\nfull_size: {self.full_size}\nchecksum: {self.checksum}\nvolume_name: {self.volume_name}"
+
+
+class RomfsExtractor(Extractor):
+    def extract(self, inpath: Path, outdir: Path):
+        with inpath.open("rb") as f:
+            header = RomFSHeader(f, outdir)
+            header.validate()
+            header.recursive_walk(header.header_end_offset, None)
+            header.dump_fs()
+
+
+def valid_checksum(content: bytes) -> int:
+    """Compute the RomFS checksum of content."""
+    total = 0
+    for i in range(0, len(content), 4):
+        total += struct.unpack(">L", content[i : (i + 4)])[0]  # noqa: E203
+        total %= MAX_UINT32
+    return total == 0
```

Comment:
This could reuse the romfs_checksum function with:

```suggestion
def valid_checksum(content: bytes) -> int:
    """Compute the RomFS checksum of content."""
    return romfs_checksum(content) == 0
```

Reply:
I've added some further refactors. Thanks for the tip